File: eclipsePlugin/org.jts.eclipse.data_1.0/templates/Common/libCSharp/lib/CSharpStateMap/Backup/FSMContext.cs
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is  State Machine Compiler(SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2004. Charles W. Rapp.
// All Rights Reserved.
// 
// Contributor(s):
//   Eitan Suez contributed examples/Ant.
//   (Name withheld) contributed the C# code generation and
//   examples/C#.
//
// RCS ID
// $Id: FSMContext.cs,v 1.8 2009/12/17 19:51:43 cwrapp Exp $
//
// statemap.java --
//
//  This package defines the FSMContext class which is inherited
//  by the smc-generated application FSM context class.
//
// CHANGE LOG
// (See the bottom of this file.)
//

using System;
using System.IO;
using System.Diagnostics;

namespace statemap
{
    // State change event. Contains the previous and new state.
    public class StateChangeEventArgs :
        EventArgs
    {
    //-----------------------------------------------------------
    // Member functions.
    //

        // Constructor.
        public StateChangeEventArgs(string fsmName,
                                    string transitionType,
                                    State previousState,
                                    State newState)
        {
            _fsmName = fsmName;
            _transitionType = transitionType;
            _previousState = previousState;
            _newState = newState;
        } // end of StateChangeEventArgs(string, State, State)

        public string FSMName()
        {
            return (_fsmName);
        } // end of FSMName()

        public string TransitionType()
        {
            return (_transitionType);
        } // end of TransitionType()

        public State PreviousState()
        {
            return (_previousState);
        } // end of PreviousState()

        public State NewState()
        {
            return (_newState);
        } // end of NewState()

    //-----------------------------------------------------------
    // Member data.
    //

        private readonly string _fsmName;
        private readonly string _transitionType;
        private readonly State _previousState;
        private readonly State _newState;
    } // end of class StateChangeEventArgs

    // Delegate declaration.
    public delegate void StateChangeEventHandler(
         object sender, StateChangeEventArgs args);

    // statemap.FSMContext --
    //
    //  Base class for the smc-generated application FSM
    //  context class.

    [Serializable]
    public abstract class FSMContext
    {
    //-----------------------------------------------------------
    // Member functions.
    //

        public FSMContext(State state)
        {
            // There is no state until the application explicitly
            // sets the initial state.
            name_ = "FSMContext";
            state_ = state;
            transition_ = "";
            previousState_ = null;
            stateStack_ = new System.Collections.Stack();
            debugFlag_ = false;
            debugStream_ = null;
        } // end of FSMContext()

        // The state change event.
        public event StateChangeEventHandler StateChange;

        // Starts the finite state machine running by executing
        // the initial state's entry action.
        public abstract void EnterStartState();

        // The finite state machine name property.
        public string Name
        {
            get
            {
                return (name_);
            }
            set
            {
                name_ = value;
            }
        } // end of Name()

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the debugFlag_, debugStream_
        // pair.
        // Used to enable debugging output
        public bool Debug
        {
            get
            {
                return debugFlag_;
            }
            set
            {
                debugFlag_ = value;
            }
        }

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the debugFlag_, debugStream_
        // pair. debugStream_ will always be null.
        // Used to set the output text writer.
        public TextWriter DebugStream
        {
            get
            {
                return debugStream_;
            }
            set
            {
                // DEPRECATED.
                // debugStream_ = value;
            }
        }

        // Is this state machine in a transition? If state is
        // null, then true; otherwise, false.
        public bool InTransition
        {
            get 
            {
                return(state_ == null ? true : false);
            }
        }

        public void SetState(State state)
        {
            StateChangeEventArgs e =
                new StateChangeEventArgs(
                    name_, "SET", state_, state);

#if TRACE
            Trace.WriteLine("ENTER STATE     : " +    state.Name);
#endif

            state_ = state;

            OnStateChange(e);

            return;
        }

        public void ClearState()
        {
            previousState_ = state_;
            state_ = null;

            return;
        }

        public State PreviousState
        {
            get
            {
                if (previousState_ != null)
                {
                    return(previousState_);
                }

                throw
                    new System.NullReferenceException(
                        "Previous state not set.");
            }
        }

        public void PushState(State state)
        {
            StateChangeEventArgs e =
                new StateChangeEventArgs(
                    name_, "PUSH", state_, state);

#if TRACE
            Trace.WriteLine("PUSH TO STATE   : " +    state.Name);
#endif

            if (state_ != null)
            {

                stateStack_.Push(state_);
            }

            state_ = state;

            OnStateChange(e);

            return;
        }

        public void PopState()
        {
            if (stateStack_.Count == 0)
            {
#if TRACE
                Trace.WriteLine("POPPING ON EMPTY STATE STACK.");
#endif

                throw new
                    System.InvalidOperationException(
                        "popping an empty state stack");
            }
            else
            {
                State nextState = (State) stateStack_.Pop();
                StateChangeEventArgs e =
                    new StateChangeEventArgs(
                        name_, "POP", state_, nextState);

                // The pop method removes the top element
                // from the stack and returns it.
                state_ = nextState;

#if TRACE
                Trace.WriteLine("POP TO STATE    : " + state_.Name);
#endif

                OnStateChange(e);
            }

            return;
        }

        public void EmptyStateStack()
        {
            stateStack_.Clear();
        }

        public string GetTransition()
        {
            return transition_;
        }

        // Release all acquired resources.
        ~FSMContext()  //TODO: Add disposable
        {
            name_ = null;
            state_ = null;
            transition_ = null;
            previousState_ = null;
            stateStack_ = null;
        }

        protected virtual void OnStateChange(StateChangeEventArgs e)
        {
            if (StateChange != null)
            {
                StateChange(this, e);
            }

            return;;
        } // end of OnStateChange(StateChangeEventArgs)

    //-----------------------------------------------------------
    // Member data
    //

        // The finite state machine's unique name.
        [NonSerialized]
        protected string name_;

        // The current state.
        [NonSerialized]
        protected State state_;

        // The current transition *name*. Used for debugging
        // purposes.
        [NonSerialized]
        protected string transition_;

        // Remember what state a transition left.
        // Do no persist the previous state because an FSM should
        // be serialized while in transition.
        [NonSerialized]
        protected State previousState_;

        // This stack is used when a push transition is taken.
        [NonSerialized]
        protected System.Collections.Stack stateStack_;

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the debugFlag_, debugStream_
        // pair. debugStream_ will always be null.
        // When this flag is set to true, this class will print
        // out debug messages.
        [NonSerialized]
        protected bool debugFlag_;

        // Write debug output to this stream.
        [NonSerialized]
        protected TextWriter debugStream_;
    } // end of class FSMContext
} // end of namespace statemap

//
// CHANGE LOG
// $Log: FSMContext.cs,v $
// Revision 1.8  2009/12/17 19:51:43  cwrapp
// Testing complete.
//
// Revision 1.7  2009/03/30 19:05:44  kgreg99
// 1. Patch for bug  #2722356. Variable stateStack_ is initialized in the constructor and not by the first "push". This prevents an unhandled exception in EmptyStateStack() without pushing a state.
//
// Revision 1.6  2009/03/01 18:20:40  cwrapp
// Preliminary v. 6.0.0 commit.
//
// Revision 1.5  2008/01/14 19:59:23  cwrapp
// Release 5.0.2 check-in.
//
// Revision 1.4  2006/09/16 15:04:28  cwrapp
// Initial v. 4.3.3 check-in.
//
// Revision 1.3  2006/06/03 19:39:25  cwrapp
// Final v. 4.3.1 check in.
//
// Revision 1.2  2006/04/22 12:45:25  cwrapp
// Version 4.3.1
//
// Revision 1.1  2005/05/28 18:44:13  cwrapp
// Updated C++, Java and Tcl libraries, added CSharp, Python and VB.
//
// Revision 1.0  2004/09/06 16:32:15  charlesr
// Initial revision
//

File: eclipsePlugin/org.jts.eclipse.data_1.0/templates/Common/libJava/lib/JavaStatemap/FSMContext.java
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is  State Machine Compiler(SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2000 - 2009. Charles W. Rapp.
// All Rights Reserved.
// 
// Contributor(s): 
//
// statemap.java --
//
//  This package defines the FSMContext class which must be
//  inherited by any Java class wanting to use an smc-generated
//  state machine.
//
// RCS ID
// $Id: FSMContext.java,v 1.14 2009/11/24 20:42:39 cwrapp Exp $
//
// CHANGE LOG
// (See the bottom of this file.)
//

package statemap;

import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.PrintStream;
import java.io.Serializable;
import java.util.EmptyStackException;

/**
 * Base class for all SMC-generated application context classes.
 * This class stores the FSM name, current and previous states,
 * the state stack, debugging information and state change
 * listeners.
 *
 * @author <a href="mailto:rapp@acm.org">Charles Rapp</a>
 */

public abstract class FSMContext
    implements Serializable
{
//---------------------------------------------------------------
// Member functions
//

    //-----------------------------------------------------------
    // Constructors.
    //

    /**
     * Creates a finite state machine context for the given
     * initial state.
     * @param initState the finite state machine's start state.
     */
    protected FSMContext(State initState)
    {
        _name = "FSMContext";
        _state = initState;
        _transition = "";
        _previousState = null;
        _stateStack = null;
        _debugFlag = false;
        _debugStream = System.err;
        _listeners = new PropertyChangeSupport(this);
    } // end of FSMContext(State)

    //
    // end of Constructors.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Abstract method declarations.
    //

    /**
     * Starts the finite state machine running by executing the
     * initial state's entry actions.
     */
    public abstract void enterStartState();

    //
    // end of Abstract method declarations.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Serializable Interface Implementation.
    //

    private void readObject(ObjectInputStream istream)
        throws IOException,
               ClassNotFoundException
    {
        istream.defaultReadObject();

        // Create an empty listeners list.
        _listeners = new PropertyChangeSupport(this);

        return;
    } // end of readObject(ObjectInputStream)

    //
    // end of Serializable Interface Implementation.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Get methods.
    //

    /**
     * Returns the FSM name.
     * @return the FSM name.
     */
    public String getName()
    {
        return (_name);
    } // end of getName()

    /**
     * When debug is set to {@code true}, the state machine
     * will print messages to the console.
     * @return {@code true} if debug output is generated.
     */
    public boolean getDebugFlag()
    {
        return(_debugFlag && _debugStream != null);
    } // end of getDebugFlag()

    /**
     * Writes the debug output to this stream.
     * @return the debug output stream.
     */
    public PrintStream getDebugStream()
    {
        return (_debugStream == null ?
                System.err :
                _debugStream);
    } // end of getDebugStream()

    /**
     * Returns {@code true} if this FSM is in a transition and
     * {@code false} otherwise.
     * @return {@code true} if this FSM is in a transition and
     * {@code false} otherwise.
     */
    public boolean isInTransition()
    {
        return(_state == null ? true : false);
    } // end of isInTransition()

    // NOTE: getState() is defined in the SMC-generated
    // FSMContext subclass.
    
    /**
     * If this FSM is in transition, then returns the previous
     * state which the last transition left.
     * @return the previous state which the current transition
     * left. May return {@code null}.
     */
    public State getPreviousState()
        throws NullPointerException
    {
        return(_previousState);
    } // end of getPreviousState()

    /**
     * If this FSM is in transition, then returns the transition
     * name. If not in trnasition, then returns an empty string.
     * @return the current transition name.
     */
    public String getTransition()
    {
        return(_transition);
    } // end of getTransition()

    //
    // end of Get methods.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Set methods.
    //

    /**
     * Sets the FSM name.
     * @param name The finite state machine name.
     */
    public void setName(String name)
    {
        if (name != null &&
            name.length() > 0 &&
            name.equals(_name) == false)
        {
            _name = name;
        }

        return;
    } // end of setName(String)

    /**
     * Turns debug output on if {@code flag} is {@code true} and
     * off if {@code flag} is {@code false}.
     * @param flag {@code true} to turn debuggin on and
     * {@code false} to turn debugging off.
     */
    public void setDebugFlag(boolean flag)
    {
        _debugFlag = flag;
        return;
    } // end of setDebugFlag(boolean)

    /**
     * Sets the debug output stream to the given value.
     * @param stream The debug output stream.
     */
    public void setDebugStream(PrintStream stream)
    {
        _debugStream = stream;
        return;
    } // end of setDebugStream(PrintStream)

    /**
     * Sets the current state to the given value.
     * @param state The current state.
     */
    public void setState(State state)
    {
        if (getDebugFlag() == true)
        {
            getDebugStream().println("ENTER STATE     : " +
                                     state.getName());
        }

        _state = state;

        // Inform any and all listeners about this state
        // change.
        _listeners.firePropertyChange(
            STATE_PROPERTY, _previousState, _state);

        return;
    } // end of setState(State)

    /**
     * Places the current state into the previous state sets
     * the current state to {@code null}.
     */
    public void clearState()
    {
        _previousState = _state;
        _state = null;

        return;
    } // end of clearState()

    /**
     * Pushes the current state on top of the state stack and
     * sets the current state to {@code state}.
     * @param state The new current state.
     * @exception NullPointerException
     * if {@code state} is {@code null}.
     */
    public void pushState(State state)
    {
        State previousState = _state;

        if (_state == null)
        {
            throw (new NullPointerException());
        }

        if (getDebugFlag() == true)
        {
            getDebugStream().println("PUSH TO STATE   : " +
                                     state.getName());
        }

        if (_stateStack == null)
        {
            _stateStack = new java.util.Stack<State>();
        }

        _stateStack.push(_state);
        _state = state;

        // Inform any and all listeners about this state
        // change.
        _listeners.firePropertyChange(
            STATE_PROPERTY, previousState, _state);

        return;
    } // end of pushState(State)

    /**
     * Sets the previous state to the current state and pops
     * the top state off the stack and places it into the
     * current state.
     * @exception EmptyStackException
     * if the state stack is empty.
     */
    public void popState()
        throws EmptyStackException
    {
        if (_stateStack == null ||
            _stateStack.isEmpty() == true)
        {
            if (getDebugFlag() == true)
            {
                getDebugStream().println(
                    "POPPING ON EMPTY STATE STACK.");
            }

            throw (new EmptyStackException());
        }
        else
        {
            State previousState = _state;

            // The pop method removes the top element
            // from the stack and returns it.
            _state = _stateStack.pop();

            if (_stateStack.isEmpty() == true)
            {
                _stateStack = null;
            }

            if (getDebugFlag() == true)
            {
                getDebugStream().println("POP TO STATE    : " +
                                      _state.getName());
            }

            // Inform any and all listeners about this state
            // change.
            _listeners.firePropertyChange(
                STATE_PROPERTY, previousState, _state);
        }

        return;
    } // end of popState()

    /**
     * Empties the state stack.
     */
    public void emptyStateStack()
    {
        if (_stateStack != null)
        {
            _stateStack.clear();
            _stateStack = null;
        }

        return;
    } // end of emptyStateStack()
	
	/**
	* Returns the name of the state on the top of the stack.
	*
	*/
	public String peakTopStackState()
	{
		if(_stateStack != null)
		{
			return _stateStack.peek().getName();
		}
		return null;	
	}
	

    //
    // end of Set methods.
    //-----------------------------------------------------------

    // The following methods allow listeners to watch this
    // finite state machine for state changes.
    // Note: if a transition does not cause a state change,
    // then no state change event is fired.

    /**
     * Adds a PropertyChangeListener to the listener list. The
     * listener is registered for state property changes only.
     * The same listener may be added more than once. For each
     * state change, the listener will be invoked the number of
     * times it was added. If {@code listener} is {@code null},
     * no exception is thrown and no action is taken.
     * @param listener The PropertyChangeListener to be added.
     */
    public void
        addStateChangeListener(PropertyChangeListener listener)
    {
        _listeners.addPropertyChangeListener(
            STATE_PROPERTY, listener);
        return;
    }

    /**
     * Removes a PropertyChangeListener for the state change
     * property. If {@code listener} was added more than once
     * to the same event source, it will be notified one less
     * time after being removed. If {@code listener} is
     * {@code null} or was never added, no exception is thrown
     * and no action is taken.
     * @param listener The PropertyChangeListener to be removed.
     */
    public void
        removeStateChangeListener(
            PropertyChangeListener listener)
    {
        _listeners.removePropertyChangeListener(
            STATE_PROPERTY, listener);
        return;
    } // end of removeStateChangeListener(PropertyChangeListener)

//---------------------------------------------------------------
// Member data
//

    /**
     * The FSM name.
     */
    transient protected String _name;

    /**
     * The current state. Will be {@code null} while in
     * transition.
     */
    transient protected State _state;

    /**
     * The current transition name. Used for debugging
     * purposes. Will be en empty string when not in
     * transition.
     */
    transient protected String _transition;

    /**
     * Stores which state a transition left. May be {@code null}.
     */
    transient protected State _previousState;

    /**
     * This stack is used to store the current state when a push
     * transition is taken.
     */
    transient protected java.util.Stack<State> _stateStack;

    /**
     * When this flag is set to {@code true}, this class will
     * print out debug messages.
     */
    transient protected boolean _debugFlag;

    /**
     * Write debug output to this stream.
     */
    transient protected PrintStream _debugStream;

    // Stores the property change listeners here.
    transient private PropertyChangeSupport _listeners;

    //-----------------------------------------------------------
    // Constants.
    //
    private static final long serialVersionUID = 0x060000L;
    private static final String STATE_PROPERTY = "State";
} // end of class FSMContext

//
// CHANGE LOG
// $Log: FSMContext.java,v $
// Revision 1.14  2009/11/24 20:42:39  cwrapp
// v. 6.0.1 update
//
// Revision 1.13  2009/09/05 15:39:20  cwrapp
// Checking in fixes for 1944542, 1983929, 2731415, 2803547 and feature 2797126.
//
// Revision 1.12  2009/03/27 09:41:07  cwrapp
// Added F. Perrad changes back in.
//
// Revision 1.11  2009/03/01 18:20:40  cwrapp
// Preliminary v. 6.0.0 commit.
//
// Revision 1.10  2008/01/14 19:59:23  cwrapp
// Release 5.0.2 check-in.
//
// Revision 1.9  2007/08/05 13:00:34  cwrapp
// Version 5.0.1 check-in. See net/sf/smc/CODE_README.txt for more information.
//
// Revision 1.8  2007/02/21 13:50:59  cwrapp
// Moved Java code to release 1.5.0
//
// Revision 1.7  2005/05/28 18:44:13  cwrapp
// Updated C++, Java and Tcl libraries, added CSharp, Python
// and VB.
//
// Revision 1.1  2005/02/21 19:03:38  charlesr
// Variable name clean up.
//
// Revision 1.0  2003/12/14 20:38:40  charlesr
// Initial revision
//
